lib/src/util.dart

```diff
@@ -96,6 +98,49 @@ String normalizeLinkLabel(String label) {
   return text;
 }
 
+///  Decodes HTML entity and numeric character references, for example decode
+/// `&#35` to `#`.
+String decodeHtmlCharacters(String input) =>
+    input.replaceAllMapped(htmlCharactersPattern, decodeHtmlCharacterFromMatch);
+
+/// Decodes HTML entity and numeric character references from the given [match].
+String decodeHtmlCharacterFromMatch(Match match) {
+  final text = match.match;
+  String? decodedText;
+
+  // Entity references, see
+  // https://spec.commonmark.org/0.30/#entity-references.
+  if (match[1] != null) {
+    decodedText = htmlEntitiesMap[text];
+  }
+
+  // Decimal numeric character references, see
+  // https://spec.commonmark.org/0.30/#decimal-numeric-character-references.
+  else if (match[2] != null) {
+    final decimalValue = int.parse(match[2]!);
+    int hexValue;
+    if (decimalValue < 1114112 && decimalValue > 1) {
+      hexValue = int.parse(decimalValue.toRadixString(16), radix: 16);
+    } else {
+      hexValue = 0xFFFd;
+    }
+
+    decodedText = String.fromCharCode(hexValue);
+  }
+
+  // Hexadecimal numeric character references, see
+  // https://spec.commonmark.org/0.30/#hexadecimal-numeric-character-references.
+  else if (match[3] != null) {
+    var hexValue = int.parse(match[3]!, radix: 16);
```

Comment:
Same here, making `match[3]` a local variable will avoid the second look up and  null check.

Reply:
Fixed